RFC-0026-logging-system.md

```diff
@@ -0,0 +1,481 @@
+slightly (for instance, if the value of some variable is included in the
+message and that value differs between two different `warnings.warn` calls),
+then both warnings are emitted. `TORCH_WARN_ONCE` does not check whether
+messages differ. But we could probably implement `torch.warn_once` in a similar
+way to how the `warnings` module filter is implemented.
+
+
+#### Info APIs
+
+Just like the error and warning APIs, the info APIs each have a variable length
+argument list, `...` in C++ and `*args` in Python. These arguments are
+concatenated into the info message.
+
+**`TORCH_LOG_INFO(...)`** - `torch.log_info(*args)`
+  - C++ info class: `c10::Info`
+  - Python warning: `torch.Info`
+  - TODO: Is there a better name than `log_info`? I didn't want to call it
+    `torch.info`, because
+    [`numpy.info`](https://numpy.org/doc/stable/reference/generated/numpy.info.html)
+    has a completely different functionality. And obviously
+    [`torch.log`](https://pytorch.org/docs/stable/generated/torch.log.html?highlight=torch%20log#torch.log)
+    is already taken.
+
+**`TORCH_LOG_INFO_WITH(info_t, ...)`** - `torch.log_info_with(info_type, *args)`
+  - C++ info class: Specified by `info_t` argument
+  - Python info class: Specified by `info_type` argument
+
+
+### Multi-process messaging APIs
+
+Currently, when running subprocesses that use PyTorch, some messages are
+emitted by every running subprocess. See
+[issue #68768](https://github.com/pytorch/pytorch/issues/68768) for specific
+examples.  Avoiding emitting duplicate messages from each subprocess by default
+would give a better user experience.
+
+In issue #68768, the duplicate messages related to `cpp_extension.load` can be
+modified to only be emitted by subprocess rank 0, simply by checking the node's
+rank first. For instance, where there is a `warnings.warn(...)`, call we can
+replace with:
+
+```python
+if rank == 0:
+    warnings.warn(...)
+```
+
+This successfully avoids duplicate warnings. A few concrete examples can be
+seen in [this draft PR](https://github.com/pytorch/pytorch/pull/79288).
+
+However, implementing the duplicate filter like this is not ideal. It would be
+better to have dedicated message system API calls for this. In the case of
+warnings, the following signature could be used:
+
+**`torch.warn_rank(my_rank, *args, warn_rank=0)`**
+  * Args:
+    - `my_rank` - Rank of the subprocess calling this function
+    - `args` - Warning message
+    - `warn_rank` - Rank that should emit the message
+  * The warning is only emitted if `my_rank == warn_rank`
+
+TODO: Add APIs for the rest of the message classes, like
+`torch.log_info_rank()`, etc.
+
+TODO: There should also be a global setting to enable emitting the duplicates.
+`torch.warn_rank` could check the setting, and if it's turned on, then it would
+emit the warning for all ranks.
+
+TODO: Should we have a `TOCH_WARN_RANK` (and others) in C++ as well? Is there
```

Comment:
This is better done by having a format string that is used to define how log lines are produced.

Reply:
@albanD, yeah I think we could make this work by default. I'm not sure if an API for checking the distributed rank currently exists in C++ (I looked briefly and didn't find it), but if not, I don't think there's any reason why we couldn't add it.

As for having all ranks send warnings/errors to rank 0, that would be interesting. I'm not sure how to implement that. Also, would sending all warnings to rank 0 

@kumpera, I'm not sure what you mean. Are you saying that you can use a format string to filter out duplicate messages from different ranks? Can you give an example?

Reply:
@kurtamohler the format of each line should be controlled by a format string.  From python's logging: https://docs.python.org/3/library/logging.html#logging.Formatter

An user dealing with logging of a distributed application could use something like: "[%(rank)] (%(time)) - %(message)" to produce a line like "[0] (2022-01-01 00:01:00) - distributed failed for some reason"

@albanD rank is available by calling `torch.distributed.get_rank()`. It's a bit more involved from C++.

On the subject of log shipping, providing a default one is a great idea as it would improve the OOTB experience.